src/remote/publication.rs

```diff
@@ -0,0 +1,1205 @@
+            ReportErrorCode::NoObjectMatchingHash => write!(f, "no_object_matching_hash"),
+            ReportErrorCode::ConsistencyProblem => write!(f, "consistency_problem"),
+            ReportErrorCode::OtherError => write!(f, "other_error"),
+        }
+    }
+}
+
+impl FromStr for ReportErrorCode {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "xml_error" => Ok(ReportErrorCode::XmlError),
+            "permission_failure" => Ok(ReportErrorCode::PermissionFailure),
+            "bad_cms_signature" => Ok(ReportErrorCode::BadCmsSignature),
+            "object_already_present" => Ok(ReportErrorCode::ObjectAlreadyPresent),
+            "no_object_present" => Ok(ReportErrorCode::NoObjectPresent),
+            "no_object_matching_hash" => Ok(ReportErrorCode::NoObjectMatchingHash),
+            "consistency_problem" => Ok(ReportErrorCode::ConsistencyProblem),
+            "other_error" => Ok(ReportErrorCode::OtherError),
+            _ => Err(Error::InvalidErrorCode(s.to_string())),
+        }
+    }
+}
+
+impl ReportErrorCode {
+    /// Provides default texts for error codes (taken from RFC).
+    fn to_text(&self) -> &str {
+        match self {
+            ReportErrorCode::XmlError => "Encountered an XML problem.",
+            ReportErrorCode::PermissionFailure => "Client does not have permission to update this URI.",
+            ReportErrorCode::BadCmsSignature => "Encountered bad CMS signature.",
+            ReportErrorCode::ObjectAlreadyPresent => "An object is already present at this URI, yet a \"hash\" attribute was not specified.",
+            ReportErrorCode::NoObjectPresent => "There is no object present at this URI, yet a \"hash\" attribute was specified.",
+            ReportErrorCode::NoObjectMatchingHash => "The \"hash\" attribute supplied does not match the \"hash\" attribute of the object at this URI.",
+            ReportErrorCode::ConsistencyProblem => "Server detected an update that looks like it will cause a consistency problem (e.g., an object was deleted, but the manifest was not updated).",
+            ReportErrorCode::OtherError => "Found some other issue."
+        }
+    }
+}
+
+
+//------------ Base64 --------------------------------------------------------
+
+/// This type contains a base64 encoded structure. The publication protocol
+/// deals with objects in their base64 encoded form.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Base64(Arc<str>);
+
+impl Base64 {
+    pub fn from_content(content: &[u8]) -> Self {
+        Base64(base64::encode(content).into())
+    }
+}
+
+impl fmt::Display for Base64 {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl Serialize for Base64 {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        self.to_string().serialize(serializer)
```

Comment:
Perhaps serialize via `str` not `String`.

Reply:
done